### PR TITLE
[xpu][test] Migrated two test files to XPU

### LINKED ISSUE #114850: [RFC] Add Intel GPU support into PyTorch CI/CD
## Motivation 
As the [[RFC] Intel GPU Upstreaming](https://github.com/pytorch/pytorch/issues/114723) mentioned, to integrate the new Intel GPU device and its associated features into PyTorch, we need to implement PyTorch CI/CD tests specifically designed for Intel GPU devices. These tests will ensure the quality of incoming pull requests and gate their acceptance accordingly. 

## Design Philosophy 
For the new part CI/CD enabling, we aim to leverage the existing PyTorch CI/CD infrastructure wherever possible. Intel GPU related test will be dispatched to Intel Develop Cloud (IDC) Instances, which provides Intel GPU hardware as self-hosted runners. Based on our understanding of current PyTorch CI/CD tests, we will divide Intel GPU related tests into several categories: pull test, inductor tests and other tests. For inductor tests, we will extend the existing inductor test workflow to accommodate Intel GPU inductor testing. And a new workflow will serve as the entry point for other tests for Intel GPU, mirroring the approach used for other devices. Overall, all Intel GPU tests follow the rules below. 
- Docker based builds & tests 
- Multiple steps both for inductor and other tests 
     + Base Docker image build on AWS instance runners which provides PyTorch build and tests environment. 
     + Wheel build on AWS instance runners directly. 
     + Tests can be sharded and dispatched on IDC instance runners. 
     
### Entrance of pull test 
For the basic build test for pull requests, we will add a new part for Intel GPU specific build in `.github/workflows/pull.yml` and triggered by each pull request.  

### Entrance of Inductor tests 
For Inductor tests, we will add a new part for Intel GPU specific tests in `.github/workflows/inductor.yml` and triggered by `ciflow/inductor`. To avoid breaking other inductor related PRs at the first stage, we plan to add a new ciflow label `ciflow/xpu` in `.github/pytorch-probot.yml`, and limit the Intel GPU inductor tests only for PR which has both `ciflow/inductor` and `ciflow/xpu`.  

### Entrance of other tests 
The Intel GPU device related remain tests will be triggered by PR with label `ciflow/xpu` or regular triggered by timer. To achieve it, we will add a new entrance workflow `.github/workflows/xpu.yml`, which like below content.
```yml
name: xpu

on:
  push:
    branches:
      - main
      - release/*
    tags:
      - ciflow/xpu/*
  workflow_dispatch:
  schedule:
    - cron: 0 0 * * *  

concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref_name }}-${{ github.ref_type == 'branch' && github.sha }}-${{ github.event_name == 'workflow_dispatch' }}-${{ github.event_name == 'schedule' }}
  cancel-in-progress: true

jobs:
  linux-jammy-xpu-py3_8-build:
    name: linux-jammy-xpu-py3.8
    uses: ./.github/workflows/_linux-build.yml
    with:
      build-environment: linux-jammy-xpu-py3.8
      docker-image-name: pytorch-linux-jammy-xpu-n-py3
      sync-tag: xpu-build
      test-matrix: |
        { include: [
          { config: "default", shard: 1, num_shards: 2, runner: "linux.idc.xpu" },
          { config: "default", shard: 2, num_shards: 2, runner: "linux.idc.xpu" },
        ]}

  linux-jammy-xpu-py3_8-test:
    name: linux-jammy-xpu-py3.8
    uses: ./.github/workflows/_xpu-test.yml
    needs: linux-jammy-xpu-py3_8-build
    with:
      build-environment: linux-jammy-xpu-py3.8
      docker-image: ${{ needs.linux-jammy-xpu-py3_8-build.outputs.docker-image }}
      test-matrix: ${{ needs.linux-jammy-xpu-py3_8-build.outputs.test-matrix }} 
```
 ### Build & Test 

Will add Intel GPU specific base image Dockerfile `.ci/docker/ubuntu-xpu/Dockerfile` and Intel GPU part into image build script `.ci/docker/build.sh` to support Intel GPU based image build on `linux.2xlarge runners`. 

For Pytorch wheel build, different with other devices, currently we need dispatch it to IDC instance runners. We will reuse `.github/workflows/_linux-build.yml` with Intel GPU specific build-environment and add Intel GPU part into Pytorch build script `.ci/pytorch/build.sh`. 

For the tests part, we will add a new Intel GPU test workflow `.github/workflows/_xpu-test.yml` and some necessary GitHub action such as `setup-xpu`, `teardown-xpu`, etc. We also will add a new part in test script `.ci/pytorch/test.sh` and a series utils scripts for Intel GPU. 

cc @seemethere @malfet @pytorch/pytorch-dev-infra @frank-wei @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### PR DESCRIPTION
**Pull request by shangerxin**
# Description
Fixes #114850, we will port test utils and schema check to Intel GPU
We could enable Intel GPU with following methods and try the best to keep the original code styles:

# Changes
1. Get device type with from accelerator and get_devtype helper method
2. Replace the requires cuda statement to device_type.
3. Add HAS_XPU and HAS GPU check to replace some of the HAS_XPU etc.

# Notify
cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela @mingfeima @sanchitintel @ashokei @jingxu10 @jerryzh168 @ipiszy @muchulee8 @aakhundov @coconutruben @gujinghui @fengyuan14 @guangyey

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -449,8 +453,8 @@ def track(x, idx):
                 # emptied at each step)
                 def hook(_unused):
                     self.assertEqual(len(stats), idx)
-                    torch.cuda.synchronize()
-                    stats.append(torch.cuda.memory_allocated())
+                    torch.get_device_module(device_type).synchronize()
```
Comment (guangyey) [inline test/test_utils.py]:
```suggestion
                    torch.accelerator.synchronize()
```

File: test/test_utils.py [hunk 2]
```diff
@@ -449,8 +453,8 @@ def track(x, idx):
...
+                    stats.append(torch.get_device_module(device_type).memory_allocated())
```
Comment (guangyey) [inline test/test_utils.py]:
```suggestion
                    stats.append(torch.accelerator.memory_allocated())
```

File: test/test_utils.py [hunk 3]
```diff
@@ -860,27 +867,27 @@ def test_get_default_device(self):
...
         try:
-            torch.set_default_device("cuda")
+            torch.set_default_device(device_type)
             self.assertEqual(torch.get_default_device(), torch.tensor([]).device)
             torch.set_default_device(None)
 
-            torch.set_default_device("cuda")
-            torch.cuda.set_device("cuda:1")
+            torch.set_default_device(device_type)
+            torch.get_device_module(device_type).set_device(f"{device_type}:1")
```
Comment (guangyey) [inline test/test_utils.py:871]:
```suggestion
            torch.accelerator.set_device_index(1)
```

Comment (shangerxin) [reply test/test_utils.py:871]: The original test logic is aim to validate from the device module. If we want to test accelerator either, should we add a new test block for accelerator?

Comment (shangerxin) [reply test/test_utils.py:871]: Added new line 877 to cover this case.

File: test/test_utils.py [hunk 4]
```diff
@@ -676,6 +681,7 @@ def test_quote_escape(self):
         for i in range(len(orig_chars)):
             self.assertEqual(self.trie.quote(orig_chars[i]), quoted_strs[i])
 
+    @unittest.skipIf(HAS_XPU, "XPU not supported hipify")
```
Comment (guangyey) [inline test/test_utils.py]: Why is not `@unittest.skipIf(Not HAS_CUDA, ...)`.

Comment (guangyey) [reply test/test_utils.py]: I think we should skip the entire `TestHipifyTrie`

File: test/test_utils.py (hunk 3 above)
Comment (guangyey) [reply test/test_utils.py:871]: Doesn't this case used to validate `get_default_device` and `set_default_device`

File: test/test_utils.py (hunk 4 above)
Comment (shangerxin) [reply test/test_utils.py]: The original case doesn't limited to run on cuda platform. So "not has_cuda" will skip the other platforms, "not has_xpu" will only skip the case for XPU.

File: test/test_utils.py (hunk 3 above)
Comment (shangerxin) [reply test/test_utils.py:871]: Yes it is. The changes are not expect to influence the original test path. We add the new line to cover the accelerator.

File: test/test_utils.py [hunk 5]
```diff
@@ -54,6 +55,9 @@
 load_tests = load_tests  # noqa: PLW0127
 
 HAS_CUDA = torch.cuda.is_available()
+HAS_XPU = torch.xpu.is_available()
+HAS_GPU = HAS_CUDA or HAS_XPU
+device_type = get_gpu_type()
```
Comment (guangyey) [inline test/test_utils.py]:
```suggestion
device_type = acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
```
It is not reasonable to make this UT depend on Inductor utils.

File: test/test_utils.py [hunk 6]
```diff
@@ -522,42 +526,43 @@ def test_infer_device_state_recursive_meta(self):
         self.assertEqual("meta", device_type)
 
     @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
-    def test_infer_device_state_recursive_multi_cuda(self):
-        # Check that no warning is issued for either cuda:0, cuda:1 or
-        # cuda:0, cuda:0 cases since they are both the same device type
+    def test_infer_device_state_recursive_multi_gpu(self):
+        # Check that no warning is issued for either gpu:0, gpu:1 or
+        # gpu:0, gpu:0 cases since they are both the same device type
+        global device_type
```
Comment (guangyey) [inline test/test_utils.py]:
```suggestion
```

File: test/test_utils.py (hunk 4 above)
Comment (guangyey) [inline test/test_utils.py]:
```suggestion
    @unittest.skipIf(not HAS_CUDA, "Hipify tests require CUDA.")
```

File: test/test_utils.py [hunk 7]
```diff
@@ -699,6 +705,7 @@ def test_prefix_words_export_trie_to_regex(self):
         expected_regex = r"(?:app(?:le)?|ban(?:ana)?)"
         self.assertEqual(regex, expected_regex)
 
+    @unittest.skipIf(HAS_XPU, "XPU not supported hipify")
```
Comment (guangyey) [inline test/test_utils.py]: ditto

File: torch/utils/checkpoint.py [hunk 8]
```diff
@@ -104,7 +104,7 @@ class DefaultDeviceType:
     to save and restore for recomputation.
     """
 
-    _default_device_type = "cuda"
+    _default_device_type = acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
```
Comment (guangyey) [inline torch/utils/checkpoint.py]: This change seems OK. May I know which test case needs this code change.

Comment (shangerxin) [reply torch/utils/checkpoint.py]: The test utils have a test case to validate the get_device_states which will reference into the checkpoint.py module.

Comment (guangyey) [reply torch/utils/checkpoint.py]: I think here perhaps introduces a problem.

Comment (guangyey) [inline torch/utils/checkpoint.py]:
```suggestion
    _default_device_type = None
```

Comment (guangyey) [reply torch/utils/checkpoint.py]:
```python
@staticmethod
    def get_device_type() -> str:
        """
        Get the current default device type for checkpointing.
        Returns:
            str: The current default device type.
        """
        if DefaultDeviceType._default_device_type:
            return DefaultDeviceType._default_device_type
        return acc.type if (acc := torch.accelerator.current_accelerator(True)) else "cpu"
```

Comment (guangyey) [reply torch/utils/checkpoint.py]: `torch.accelerator.current_accelerator(True)` perhaps change at the post init process espacially working with out-of-tree backend. So we should not initialize it at import time

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@guangyey land it when you're ready, I'm just endorsing

**Comment by guangyey** [comment]
> land it when you're ready, I'm just endorsing

Thanks very much~

File: test/test_utils.py (hunk 4 above)
Comment (shangerxin) [reply test/test_utils.py]: Hi @guangyey , The original code is not limit to run on cuda. Should we use not HAS_CUDA to limit the case only run on cuda platform?

File: torch/utils/checkpoint.py (hunk 8 above)
Comment (shangerxin) [reply torch/utils/checkpoint.py]: We can assign the value _default_device_type to None. When gget the property if the value is None then assign the value to default acc type.

File: test/test_utils.py (hunk 4 above)
Comment (guangyey) [reply test/test_utils.py]: You are right!

**Review by guangyey — APPROVED** [review]
Thanks for update!

**Comment by guangyey** [comment]
Please fix the failure.

**Comment by shangerxin** [comment]
Working on it.

File: test/test_schema_check.py [hunk 9]
```diff
@@ -18,6 +19,7 @@
 pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
 sys.path.append(pytorch_test_dir)
 
+HAS_XPU = torch.xpu.is_available()
```
Comment (guangyey) [inline test/test_schema_check.py]: Please use the already defined `TEST_XPU` in torch/testing/_internal/common_utils.py

Comment (shangerxin) [reply test/test_schema_check.py]: Done.

**Comment by shangerxin** [comment]
Hi @guangyey ,
Need your helps to review the code, the skip test cases logic is updated. 
Thank you.

File: test/test_schema_check.py [hunk 10]
```diff
@@ -496,19 +496,67 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         with SchemaInfoBindTestMode(self) as schemaInfoCheck:
             x.add(x)
 
+
 class TestSchemaCheckModeOpInfo(JitTestCase):
+    _xpu_not_supported_cases = ('test_schema_correctness_amax_xpu_uint16',
```
Comment (daisyden) [inline test/test_schema_check.py]: Could you refer to https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_opinfo.py#L189 to define the skip list in a more simple way?

File: torch/testing/_internal/common_utils.py [hunk 11]
```diff
@@ -107,6 +107,9 @@
 NAVI_ARCH = ("gfx1030", "gfx1100", "gfx1101", "gfx1200", "gfx1201")
 NAVI3_ARCH = ("gfx1100", "gfx1101")
 NAVI4_ARCH = ("gfx1200", "gfx1201")
+HAS_XPU = torch.xpu.is_available()
```
Comment (daisyden) [inline torch/testing/_internal/common_utils.py]: Why not use TEST_XPU and TEST_CUDA?

File: test/test_utils.py [hunk 12]
```diff
@@ -435,10 +437,10 @@ def run_fn2(tensor1, tensor2):
             out = checkpoint(run_fn2, input_var, input_var2, use_reentrant=True)
             out.sum().backward()
 
-    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA")
+    @unittest.skipIf(not HAS_GPU, "Test requires GPU")
```
Comment (daisyden) [inline test/test_utils.py]: Could you try TEST_ACCELERATOR?

File: test/test_schema_check.py (hunk 10 above)
Comment (shangerxin) [reply test/test_schema_check.py]: The skip list is only relative to this test class. It will not be referenced in other place.

Comment (guangyey) [reply test/test_schema_check.py]: This code change is so big. I think we'd better restore the code change and land the others. Then file an issue to torch-xpu-ops, I think the op listed here should pass the schema check.

File: torch/testing/_internal/common_utils.py (hunk 11 above)
Comment (guangyey) [reply torch/testing/_internal/common_utils.py]: I think Daisy is right. We better use the variable already defined.

File: test/test_schema_check.py (hunk 10 above)
Comment (shangerxin) [reply test/test_schema_check.py]: OK.

File: torch/testing/_internal/common_utils.py (hunk 11 above)
Comment (shangerxin) [reply torch/testing/_internal/common_utils.py]: The TEST_XPU is a new added flag in main branch. I will pick it. Thanks.

File: test/test_schema_check.py (hunk 10 above)
Comment (shangerxin) [reply test/test_schema_check.py]: The lint issues are fixed in local environment.

File: torch/testing/_internal/common_utils.py [hunk 13]
```diff
@@ -108,6 +108,7 @@
 NAVI3_ARCH = ("gfx1100", "gfx1101")
 NAVI4_ARCH = ("gfx1200", "gfx1201")
 
+
```
Comment (daisyden) [inline torch/testing/_internal/common_utils.py]: no need this newline, right？

**Review by guangyey — APPROVED** [review]

**Comment by shangerxin** [comment]
The job failed due to 

```
The job failed due to a recurring backend compiler exception in PyTorch Inductor for XPU, specifically in the compile_fx_wrapper backend when handling the operator
```
and 
```
test timeout
```
